File: DotNet/PopulationFitness/PopulationFitness/Models/Individual.cs
using PopulationFitness.Models.Genes;
using System;

namespace PopulationFitness.Models
{
    /**
     * Created by pete.callaghan on 03/07/2017.
     */
    public class Individual
    {
        private readonly Epoch _epoch;

        public readonly int BirthYear;

        public readonly IGenes Genes;

        public Individual(Epoch epoch, int birth_year)
        {
            _epoch = epoch;
            BirthYear = birth_year;
            Genes = epoch.Config.GenesFactory.Build(epoch.Config);
        }

        public int Age(int current_year)
        {
            return Math.Abs(current_year - BirthYear);
        }

        public bool IsReadyToDie(int current_year)
        {
            return Age(current_year) >= _epoch.MaxAge();
        }

        public bool CanBreed(int current_year)
        {
            int age = Age(current_year);
            return age >= _epoch.Config.MinBreedingAge && age <= _epoch.MaxBreedingAge();
        }

        public int InheritFromParents(Individual mother, Individual father)
        {
            return Genes.InheritFrom(mother.Genes, father.Genes);
        }
    }
}

File: Scala/PopulationFitness/src/main/scala/uk/edu/populationfitness/models/Individual.scala
package uk.edu.populationfitness.models

import uk.edu.populationfitness.models.genes.Fitness
import uk.edu.populationfitness.models.genes.bitset.BitSetGenes

object Individual{
  def inheritFromParents(epoch: Epoch, birthYear: Int, mother: Individual, father: Individual): Individual = {
    new Individual(epoch, birthYear, BitSetGenes.inheritFrom(mother.genes, father.genes))
  }

  def apply(epoch: Epoch, birthYear: Int): Individual = new Individual(epoch, birthYear, BitSetGenes.buildRandom(epoch.config))
}

class Individual private[models] (val epoch: Epoch, val birthYear: Int, val genes: BitSetGenes) {

  private val _fitness: Fitness = uk.edu.populationfitness.models.genes.fitness.Fitness(epoch.config)

  def fitness: Double = _fitness(genes)

  def age(currentYear: Int): Int = math.abs(currentYear - birthYear)

  def isReadyToDie(currentYear: Int): Boolean = age(currentYear) >= epoch.maxAge

  def canBreed(currentYear: Int): Boolean = {
    val a = age(currentYear)
    (a >= epoch.config.minBreedingAge) && (a <= epoch.maxBreedingAge)
  }
}
